snowflake/src/test/java/org/apache/iceberg/snowflake/FakeSnowflakeClient.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.snowflake;
+
+import java.util.List;
+import java.util.Map;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.jdbc.UncheckedSQLException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.snowflake.entities.SnowflakeSchema;
+import org.apache.iceberg.snowflake.entities.SnowflakeTable;
+import org.apache.iceberg.snowflake.entities.SnowflakeTableMetadata;
+
+public class FakeSnowflakeClient implements SnowflakeClient {
```

Comment:
Right, my motivation for the refactor was specifically to make SnowflakeCatalogTest agnostic of the client implementation - that's the goal to make the unittest well-contained. We could always also add a unittest that's more of an integration test if we have a good way to mock out the lower JDBC communication, but ultimately if we supply a fake SnowflakeClient anyways that doesn't call into JdbcSnowflakeClient, that client should avoid using JDBC under the hood.

If we can inject something into a real JdbcSnowflakeClient that achieves what we want then we can inject that real JdbcSnowflakeClient into a real SnowflakeCatalog for a more end-to-end test.

Otherwise, for single-class unittests we can keep them clean and self-contained.